R/fa_i.R

```diff
@@ -24,19 +41,54 @@
 #' @export
 fa_i <- function(name,
                  class = NULL,
+                 htmlDependency = NULL,
                  ...) {
 
   prefix <- "fa"
   iconClass <- ""
 
-  if (!is.null(name)) {
-    prefix_class <- prefix
-    if (prefix_class == "fa" && name %in% font_awesome_brands) {
-      prefix_class <- "fab"
-    }
-    iconClass <- paste0(prefix_class, " ", prefix, "-", name)
+  # Determine if the `name` is a Font Awesome v4
+  # icon name and provide a warning
+  if (name %in% fa_tbl$v4_name && !(name %in% fa_tbl$name)) {
+
+    # Obtain the version 5 `name` and `full_name`
+    # for messaging purposes
+    v5_name <- fa_tbl[fa_tbl$v4_name == name, ][1, "name"]
+    v5_name_full <- fa_tbl[fa_tbl$v4_name == name, ][1, "full_name"]
+
+    # Warn that the v4 icon name should be changed to a v5 one
+    warning(
+      "The `name` provided ('", name ,"') is deprecated in Font Awesome v5:\n",
+      "* please consider using '", v5_name, "' or '", v5_name_full, "' instead",
+      call. = FALSE
+    )
+  }
+
+  # Provide a warning if the icon name can't be resolved, but
+  # only do this if the user isn't supplying a custom `html-dependency`
+  # (which is the common scenario)
+  if (is.null(htmlDependency) &&
+      !(name %in% fa_tbl$full_name) &&
+      !(name %in% fa_tbl$name) &&
+      !(name %in% fa_tbl$v4_name)
+  ) {
+    warning("This Font Awesome icon ('", name, "') does not exist", call. = FALSE)
+    make_browsable <- FALSE
+  } else {
+    make_browsable <- TRUE
+  }
+
+  prefix_class <- prefix
+
+  font_awesome_brands <-
+    unique(fa_tbl$name[grepl("fab ", fa_tbl$full_name)])
```

Comment:
This looked like it could be slow, so I profiled it. Running `fa_i('calendar')` 1000 times takes 700ms on my computer, and the majority of it is spent on this line:
![image](https://user-images.githubusercontent.com/86978/109822839-f528d100-7bfc-11eb-8ceb-8d6363d7a57f.png)

We should be able to speed this up. Perhaps precompute this data structure?

Reply:
I moved this computation over to the data processing script and now the vector is available as internal data. Now it's just retrieval (no more `grepl()`, subsetting, and using `unique()`).